libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,119 @@
+
+	// Function name and a pair of boolean masks, the first corresponds to parameters and the second
+	// corresponds to returnVariables.
+	//
+	// For the first vector in the pair, a value `false` at index `i` indicates that the function
+	// argument at index `i` in `FunctionDefinition::parameters` is unused inside the function body.
+	//
+	// Similarly for the second vector in the pair, a value `false` at index `i` indicates that the
+	// return parameter at index `i` in `FunctionDefinition::returnVariables` is unused inside
+	// function body.
+	map<YulString, pair<vector<bool>, vector<bool>>> usedParametersAndReturnVariables;
+
+	// Step 1 of UnusedFunctionParameterPruner: Find functions whose parameters (both arguments and
+	// return-parameters) are not used in its body.
+	for (auto const& statement: _ast.statements)
+		if (holds_alternative<FunctionDefinition>(statement))
+		{
+			FunctionDefinition const& f = std::get<FunctionDefinition>(statement);
+
+			if (tooSimpleToBePruned(f) || boost::algorithm::all_of(f.parameters + f.returnVariables, used))
+				continue;
+
+			usedParametersAndReturnVariables[f.name] = {
+				applyMap(f.parameters, used),
+				applyMap(f.returnVariables, used)
+			};
+		}
+
+	set<YulString> functionNamesToFree = util::keys(usedParametersAndReturnVariables);
+
+	// Step 2 of UnusedFunctionParameterPruner: Renames the function and replaces all references to
+	// the function, say `f`, by its new name, say `f_1`.
+	NameDisplacer replace{_context.dispenser, functionNamesToFree};
+	replace(_ast);
+
+	// Inverse-Map of the above translations. In the above example, this will store an element with
+	// key `f_1` and value `f`.
+	std::map<YulString, YulString> newToOriginalNames = invertMap(replace.translations());
+
+	// Step 3 of UnusedFunctionParameterPruner: introduce a new function in the block with body of
+	// the old one. Replace the body of the old one with a function call to the new one with reduced
+	// parameters.
+	//
+	// For example: introduce a new 'linking' function `f` with the same the body as `f_1`, but with
+	// reduced parameters, i.e., `function f() -> y { y := 1 }`. Now replace the body of `f_1` with
+	// a call to `f`, i.e., `f_1(x) -> y { y := f() }`.
+	iterateReplacing(_ast.statements, [&](Statement& _s) -> optional<vector<Statement>> {
+		if (holds_alternative<FunctionDefinition>(_s))
+		{
+			FunctionDefinition& originalFunctionWithNewName = get<FunctionDefinition>(_s);
+			if (newToOriginalNames.count(originalFunctionWithNewName.name))
+			{
+				YulString originalName = newToOriginalNames.at(originalFunctionWithNewName.name);
+
+				// After this, the name of `originalFunctionWithNewName` will be changed back to
+				// original, i.e., `f_1 -> f`. Its parameters (both arguments and return-variables)
+				// will be pruned. The function returned will have the new name, i.e., `f_1`, and
+				// becomes the 'linking' function.
+				FunctionDefinition linkingFunction = createLinkingFunction(
```

Comment:
I would find it more natural if `createLinkingFunction` did not modify the original function and really would just create the linking function. As far as I can see, the modifications on the original function would just be three lines of code that can easily be pulled out - is that correct?

Reply:
I moved the modifications for the original function to be outside now.